### PR TITLE
Feature/terraform import

### PR DESCRIPTION
**Pull request by aliasgharmhowwala**

### CODE REVIEW

**Comment by aliasgharmhowwala** [comment]
# Testing Approach
## Manual Testing 
- Created the plugin binary with version v0.2.0
- Copied it to plugins directory `~/.terraform.d/plugins/`.
- Initialized the plugin with `terraform init` command.
- Next, generated a plan using `terraform plan`
- Created the resources on the test environment using `terraform apply` command.
- Removed the state file `terraform.tfstate` so that we can test the import feature.
- Then imported each of the resources present in main.tf using `terraform import` command.
- Finally, destroyed the resources using `terraform destroy` command.
- Tried the above steps with both terraform 0.11 and 0.12 for backward and forward compatibility.

Below are the results/output of the testing:

[Terraform 11 import feature testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305434/Terraform.11.import.feature.testing.txt)
[Terraform 12 import feature testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305435/Terraform.12.import.feature.testing.txt)

## Integration Testing
- Below is the results/output of integration testing:
[Integration testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305530/Integration.testing.txt)

## Unit Testing
- Below is the results/output of unit testing:
[unit testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305065/unit.testing.txt)

File: .github/FEATURE_REQUEST.md
Comment (kbhandari) [inline .github/FEATURE_REQUEST.md]: From where did u got this file from, please get it reviewed by Ben. Same thing holds true for other .md files

File: CHANGELOG.md
Comment (kbhandari) [inline CHANGELOG.md]: May be replace it with something like "improved test coverage"... Ben can help more, please check with him

File: CHANGELOG.md
Comment (kbhandari) [inline CHANGELOG.md]: Define it? I think better way to put this is something like: make the repo compatible with latest ultradns-sdk-go plugin.

File: CHANGELOG.md
Comment (kbhandari) [inline CHANGELOG.md]: I think we should only mention feature bugs here.... please get the whole file reviewed by Ben.

File: ultradns/common.go
Comment (kbhandari) [inline ultradns/common.go]: Please replace ur personal fork reference from here.

File: ultradns/common.go
Comment (kbhandari) [inline ultradns/common.go]: You absolutely sure about this change? looks like it can have some major side effects. Please test thoroughly.

File: ultradns/common_test.go
Comment (kbhandari) [inline ultradns/common_test.go]: No personal for reference please.

File: ultradns/config.go
Comment (kbhandari) [inline ultradns/config.go]: Please remove personal for reference from everywhere.

File: ultradns/resource_ultradns_probe_http.go
Comment (kbhandari) [inline ultradns/resource_ultradns_probe_http.go]: No personal fork reference please.

File: ultradns/resource_ultradns_probe_http.go
Comment (kbhandari) [inline ultradns/resource_ultradns_probe_http.go:143]: Please make sure that these colons separators doesn't collide with any actual colons.

**Review by kbhandari — COMMENTED** [review]
> # Testing Approach
> ## Manual Testing
> 
>     * Created the plugin binary with version v0.2.0
> 
>     * Copied it to plugins directory `~/.terraform.d/plugins/`.
> 
>     * Initialized the plugin with `terraform init` command.
> 
>     * Next, generated a plan using `terraform plan`
> 
>     * Created the resources on the test environment using `terraform apply` command.
> 
>     * Removed the state file `terraform.tfstate` so that we can test the import feature.
> 
>     * Then imported each of the resources present in main.tf using `terraform import` command.
> 
>     * Finally, destroyed the resources using `terraform destroy` command.
> 
>     * Tried the above steps with both terraform 0.11 and 0.12 for backward and forward compatibility.
> 
> 
> Below are the results/output of the testing:
> 
> [Terraform 11 import feature testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305434/Terraform.11.import.feature.testing.txt)
> [Terraform 12 import feature testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305435/Terraform.12.import.feature.testing.txt)
> ## Integration Testing
> 
>     * Below is the results/output of integration testing:
>       [Integration testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305530/Integration.testing.txt)
> 
> 
> ## Unit Testing
> 
>     * Below is the results/output of unit testing:
>       [unit testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305065/unit.testing.txt)

Did you tested the import for some specific resource? If not, please test it as well.

File: .github/FEATURE_REQUEST.md
Comment (aliasgharmhowwala) [reply .github/FEATURE_REQUEST.md]: Got this files from the OSS Page, but then again will get it reviewed by Ben

**Comment by aliasgharmhowwala** [comment]
> > # Testing Approach
> > ## Manual Testing
> > ```
> > * Created the plugin binary with version v0.2.0
> > 
> > * Copied it to plugins directory `~/.terraform.d/plugins/`.
> > 
> > * Initialized the plugin with `terraform init` command.
> > 
> > * Next, generated a plan using `terraform plan`
> > 
> > * Created the resources on the test environment using `terraform apply` command.
> > 
> > * Removed the state file `terraform.tfstate` so that we can test the import feature.
> > 
> > * Then imported each of the resources present in main.tf using `terraform import` command.
> > 
> > * Finally, destroyed the resources using `terraform destroy` command.
> > 
> > * Tried the above steps with both terraform 0.11 and 0.12 for backward and forward compatibility.
> > ```
> > 
> > 
> > Below are the results/output of the testing:
> > [Terraform 11 import feature testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305434/Terraform.11.import.feature.testing.txt)
> > [Terraform 12 import feature testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305435/Terraform.12.import.feature.testing.txt)
> > ## Integration Testing
> > ```
> > * Below is the results/output of integration testing:
> >   [Integration testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305530/Integration.testing.txt)
> > ```
> > 
> > 
> > ## Unit Testing
> > ```
> > * Below is the results/output of unit testing:
> >   [unit testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305065/unit.testing.txt)
> > ```
> 
> Did you tested the import for some specific resource? If not, please test it as well.

Yup we have tested the import feature on the individual resources for that we can refer to the files 
 [Terraform 11 import feature testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305434/Terraform.11.import.feature.testing.txt)
[Terraform 12 import feature testing.txt](https://github.com/terraform-providers/terraform-provider-ultradns/files/5305435/Terraform.12.import.feature.testing.txt)

File: CHANGELOG.md
Comment (aliasgharmhowwala) [reply CHANGELOG.md]: Done got Reviews from Ben and again sent him for final review

File: CHANGELOG.md
Comment (aliasgharmhowwala) [reply CHANGELOG.md]: Done got Reviews from Ben and again sent him for final review

File: CHANGELOG.md
Comment (aliasgharmhowwala) [reply CHANGELOG.md]: Done got Reviews from Ben and again sent him for final review

File: ultradns/common.go
Comment (aliasgharmhowwala) [reply ultradns/common.go]: yup will do it once we merge the sdk plugin to master in ultradns project

File: ultradns/common.go
Comment (aliasgharmhowwala) [reply ultradns/common.go]: yup we tested it thoroughly and we are sure about the change

File: ultradns/common_test.go
Comment (aliasgharmhowwala) [reply ultradns/common_test.go]: yup will do it once we merge the sdk plugin to master in ultradns project

File: ultradns/config.go
Comment (aliasgharmhowwala) [reply ultradns/config.go]: yup will do it once we merge the sdk plugin to master in ultradns project

File: ultradns/resource_ultradns_probe_http.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_probe_http.go:143]: yup it won't collide with actual colons as per the RFC there will be no `:` in the zone and resource record names

File: ultradns/resource_ultradns_probe_http.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_probe_http.go]: yup will do it once we merge the sdk plugin to master in ultradns project

File: vendor/cloud.google.com/go/CHANGES.md
Comment (mushkoor) [inline vendor/cloud.google.com/go/CHANGES.md]: Vendor libraries should not be part of our code base. These are dependencies that should be downloaded during build.

FYI: @kbhandari

File: .github/SUPPORT.md
Comment (mushkoor) [inline .github/SUPPORT.md:8]: Please add SECURITY.md

File: acceptence_test/main.tf
Comment (mushkoor) [inline acceptence_test/main.tf:289]: Why ttl was removed?

File: ultradns/common.go
Comment (mushkoor) [inline ultradns/common.go]: The format of RRSet ID has been changed. Is it required for import feature?

File: ultradns/resource_ultradns_dirpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_dirpool.go:350]: Is this log going to add any value to the user of this plugin?

File: ultradns/resource_ultradns_dirpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_dirpool.go:384]: Why logging has been added here? Is it going to add some value to plugin user?

File: ultradns/resource_ultradns_dirpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_dirpool.go:396]: is typ a typo or it is intentional.

File: ultradns/resource_ultradns_dirpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_dirpool.go:431]: Is this log going to add some value to the user of this plugin?

File: ultradns/resource_ultradns_dirpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_dirpool.go:445]: In what case unmarshalling fail? Will user of this plugin be able to take corrective action based on the errror message?

File: ultradns/resource_ultradns_dirpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_dirpool.go:448]: There is no spacing before {. Please check all new code with golang guidelines.

File: ultradns/resource_ultradns_dirpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_dirpool.go:456]: When setting no_reponse field will fail? Will this error message be visible to the user of this plugin?
If yes, noResponseBlock is the internal variable name used in our code. The user is not aware of noResponseBlock.

File: ultradns/resource_ultradns_dirpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_dirpool.go:493]: When makeIPInfo will fail? Will this error message be visible to the user of this plugin. Will this message help user to understand the issue and take corrective actions.

File: ultradns/resource_ultradns_dirpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_dirpool.go:500]: When makeGeoInfo will fail? Will this error message be visible to the user of this plugin. Will this message help user to understand the issue and take corrective actions.

File: ultradns/resource_ultradns_dirpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_dirpool.go]: Why we have put extra lines before and after errors.New

File: ultradns/resource_ultradns_probe_http.go
Comment (mushkoor) [inline ultradns/resource_ultradns_probe_http.go]: We have extra lines before and after errors.New(

File: ultradns/resource_ultradns_probe_ping.go
Comment (mushkoor) [inline ultradns/resource_ultradns_probe_ping.go:164]: I assume setting p.ID twice has been done for backward compatibility.

File: ultradns/resource_ultradns_probe_ping.go
Comment (mushkoor) [inline ultradns/resource_ultradns_probe_ping.go:191]: Is logging going to add any value to the end user of this plugin? Will these logs be visible to end user?

File: ultradns/resource_ultradns_rdpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_rdpool.go:144]: Why logging is required? Will it add value to end user?

File: ultradns/resource_ultradns_rdpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_rdpool.go:213]: Why logging is required? Will it add value to end user?

File: ultradns/resource_ultradns_record.go
Comment (mushkoor) [inline ultradns/resource_ultradns_record.go]: Please remove commented code.

File: ultradns/resource_ultradns_record.go
Comment (mushkoor) [inline ultradns/resource_ultradns_record.go]: Will this log be visible to end user? Will it add any value to the end user? There is no contextual information in the log.

File: ultradns/resource_ultradns_record.go
Comment (mushkoor) [inline ultradns/resource_ultradns_record.go:177]: Please review all the logs from contextual information and usefulness from end user's point of view.

File: ultradns/resource_ultradns_record.go
Comment (mushkoor) [inline ultradns/resource_ultradns_record.go]: Is ID provided by end user? It is calculated by the plugin itself.

Is message "Wrong ID please provide proper ID in format name:zone " relevant from end user's perspective.

File: ultradns/resource_ultradns_tcpool.go
Comment (mushkoor) [inline ultradns/resource_ultradns_tcpool.go]: I can see this code is repeated in almost every resource. It can be moved to a common place.

File: ultradns/provider_test.go
Comment (mushkoor) [inline ultradns/provider_test.go:93]: I don't see the acceptance test case for import feature.

File: ultradns/resource_ultradns_dirpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go:350]: Just for Debugging purpose

Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go:351]: will remove it

File: ultradns/resource_ultradns_record.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_record.go]: No it won't be visible at user end. it will be only visible while doing the integration testing

File: ultradns/resource_ultradns_record.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_record.go]: Yup the ID will be given by the end user while importing the resource, so if they provide wrong ID then we won't be able to import the resource

File: ultradns/common.go
Comment (aliasgharmhowwala) [reply ultradns/common.go]: Yup it is required for the terraform import and basically used for ID Generation

File: ultradns/provider_test.go
Comment (aliasgharmhowwala) [reply ultradns/provider_test.go:93]: Provider is not a resource it can't be imported

File: ultradns/resource_ultradns_dirpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go:384]: It is for Debugging purpose for engineer , it won't be visible to the end-user

File: ultradns/resource_ultradns_dirpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go:431]: It will give the engineer a debugging point, won't be visible when end user is using it

File: ultradns/resource_ultradns_dirpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go:445]: The unmarshalling will fail in case we got some extra fields coming in from API when there is a DTO upgrade and it will give the error and yup engineer will be able to take corrective actions cause i was also able to debug using the same.

File: ultradns/resource_ultradns_dirpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go:456]: yup it can fail when the fields are not coming in from API as expected in case of DTO Upgrade so when the user will try to import it  it won't be importable in that case the import will fail and yup it won't make sense to the user in that case they will open a ticket and we will have to take a look into it

File: ultradns/resource_ultradns_dirpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go:493]: This messages are intended for us if end user see's it they will have to raise a ticket for us to look into it they won't be able to take corrective action

File: ultradns/resource_ultradns_dirpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go:500]: This messages are intended for us if end user see's it they will have to raise a ticket for us to look into it they won't be able to take corrective action

File: ultradns/resource_ultradns_probe_ping.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_probe_ping.go:164]: It is set twice because the ID parameter needs to be generated in a format that we could later use it to import it back and plus the second time it is set to send actual ID in the URL while creating the resource at API end

File: ultradns/resource_ultradns_probe_ping.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_probe_ping.go:191]: It is valuable at engineer's view and will not be visible to customer

File: ultradns/resource_ultradns_rdpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_rdpool.go:144]: It will not be visible to end user, will only be visible while doing integration testing when developer does it

File: ultradns/resource_ultradns_rdpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_rdpool.go:213]: It will not be visible to end user, will only be visible while doing integration testing when developer does it

File: ultradns/resource_ultradns_record.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_record.go]: yup will remove it

File: ultradns/resource_ultradns_record.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_record.go:177]: It will not be visible to end user, will only be visible while doing integration testing when developer does it

File: ultradns/resource_ultradns_tcpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_tcpool.go]: for some of the resources we will need special handling for setting up the type for the resources, so i think it will be more manageable on resource basis and also recommended by terraform

File: acceptence_test/main.tf
Comment (aliasgharmhowwala) [reply acceptence_test/main.tf:289]: because it will be a part of RDATAINFO so TTL cannot be provided over here

File: vendor/cloud.google.com/go/CHANGES.md
Comment (aliasgharmhowwala) [reply vendor/cloud.google.com/go/CHANGES.md]: Removed the libs

File: ultradns/resource_ultradns_tcpool.go
Comment (mushkoor) [reply ultradns/resource_ultradns_tcpool.go]: We need to remove duplicate code. Please remove. Wherever we need special handling, do special handling in that resource itself.

File: .github/SUPPORT.md
Comment (aliasgharmhowwala) [reply .github/SUPPORT.md:8]: Done added the file

File: ultradns/resource_ultradns_dirpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go:396]: It is intentional since type is a keyword in go

File: ultradns/resource_ultradns_dirpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go:448]: all the files are already formatted by go formatter

File: ultradns/resource_ultradns_dirpool.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_dirpool.go]: Removed them as well

File: ultradns/resource_ultradns_probe_http.go
Comment (aliasgharmhowwala) [reply ultradns/resource_ultradns_probe_http.go]: removed them as well

**Comment by vpavlyuk** [comment]
`Using the provider` section is not filled in on the github project page.

**Comment by vpavlyuk** [comment]
Is business/legal OK with the Mozilla license? I know we're OK with Apache2 & MIT, not sure about Mozilla. If confirmed, please include Mozilla into `UltraDNS OSS Process & Guidelines` document.

**Comment by aliasgharmhowwala** [comment]
> `Using the provider` section is not filled in on the github project page.

We have incorporated the recommended changes and also have reviewed it once with Ben

**Review by mushkoor — APPROVED** [review]

**Pull request merged**
